src/Microsoft.FeatureManagement/Targeting/ContextualTargetingFeatureVariantAssigner.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+//
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Options;
+using Microsoft.FeatureManagement.FeatureFilters;
+using Microsoft.FeatureManagement.Targeting;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.FeatureManagement
+{
+    /// <summary>
+    /// A feature variant assigner that can be used to assign a variant based on targeted audiences.
+    /// </summary>
+    [FilterAlias(Alias)]
+    public class ContextualTargetingFeatureVariantAssigner : IContextualFeatureVariantAssigner<ITargetingContext>
+    {
+        private const string Alias = "Microsoft.Targeting";
+        private readonly TargetingEvaluationOptions _options;
+
+        /// <summary>
+        /// Creates a targeting contextual feature filter.
+        /// </summary>
+        /// <param name="options">Options controlling the behavior of the targeting evaluation performed by the filter.</param>
+        public ContextualTargetingFeatureVariantAssigner(IOptions<TargetingEvaluationOptions> options)
+        {
+            _options = options?.Value ?? throw new ArgumentNullException(nameof(options));
+        }
+
+        /// <summary>
+        /// Assigns one of the variants configured for a feature based off the provided targeting context.
+        /// </summary>
+        /// <param name="variantAssignmentContext">Contextual information available for use during the assignment process.</param>
+        /// <param name="targetingContext">The targeting context used to determine which variant should be assigned.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
+        /// <returns></returns>
+        public ValueTask<FeatureVariant> AssignVariantAsync(FeatureVariantAssignmentContext variantAssignmentContext, ITargetingContext targetingContext, CancellationToken cancellationToken)
+        {
+            FeatureDefinition featureDefinition = variantAssignmentContext.FeatureDefinition;
+
+            if (featureDefinition == null)
+            {
+                return new ValueTask<FeatureVariant>((FeatureVariant)null);
+            }
+
+            FeatureVariant variant = null;
+
+            FeatureVariant defaultVariant = null;
+
+            double cumulativePercentage = 0;
+
+            var cumulativeGroups = new Dictionary<string, double>();
+
+            if (featureDefinition.Variants != null)
+            {
+                foreach (FeatureVariant v in featureDefinition.Variants)
+                {
+                    if (defaultVariant == null && v.Default)
```

Comment:
Should we throw if the default is true for more than one variant?

Reply:
Yes. We had another comment with you and Avani. It looks like we all agree to handle that possibility in `FeatureManager`.

Reply:
We now throw if there is no default, or if there is more than 1 default.